static/fetch_lib.sh

```diff
@@ -0,0 +1,42 @@
+#!/bin/bash
+# shellcheck disable=2034,2059
+true
+# see https://github.com/koalaman/shellcheck/wiki/Directive
+
+# Functions
+lib_error_message() {
+    msg_box "You don't have internet and the local lib isn't available. Hence you cannot run this script."
+    exit 1
+}
+
+download_cache_lib() {
+    # Cache the lib for half an hour
+    if ! [ -f /var/scripts/lib.sh ] || test "$(find /var/scripts/lib.sh -mmin +30)"
+    then
+        rm -f /var/scripts/lib.sh
+        mkdir -p /var/scripts
+        if ! curl -so /var/scripts/lib.sh https://raw.githubusercontent.com/nextcloud/vm/master/lib.sh
+        then
+            lib_error_message
+        fi
+    fi
+}
+
+# Run the script
+if [ -f /var/scripts/nextcloud-startup-script.sh ] && ! [ -f /var/scripts/you-can-not-run-the-startup-script-several-times ]
```

Comment:
Does that make sense to you? (I just added the second condition)

Reply:
I don't understand what that does. Why do we need to check if the startup script?

fetch_lib should only fetch either the local lib or curl if we have internet.

Reply:
because we have to differentiate between the startup-script and all other cases.
Or do you just want to check always if the internet works before continuing with any script?
(doesn't make really sense, imo since many scripts get downloaded before they get executed.)

Reply:
maybe we should talk about what it should do:
Currently it tests the internet connection for every script during the startup-script, as long as the you-can-not-run-the-startup-script-twice file isn't there.
If one of these cases aren't met, it always downloads the new library from github, if the local library doesn't exist or if the local library is older than 30 minutes. If one of those cases is met, it will first remove the local lib and will try to download the new lib, and will report and exit if this download hasn't succeeded.

Reply:
The idea here is to avoid having to type like 20 lines

```
# T&M Hansson IT AB © - 2020, https://www.hanssonit.se/

# If we have internet, then use the latest variables from the lib remote file
if printf "Testing internet connection..." && ping github.com -c 2
then
# shellcheck disable=2034,2059
true
SCRIPT_NAME="Server configuration"
# shellcheck source=lib.sh
. <(curl -sL https://raw.githubusercontent.com/nextcloud/vm/master/lib.sh)
# Use local lib file in case there is no internet connection
elif [ -f /var/scripts/lib.sh ]
then
# shellcheck disable=2034,2059
true
# shellcheck source=lib.sh
source /var/scripts/lib.sh
else
    printf "You don't seem to have a working internet connection, and /var/scripts/lib.sh is missing so you can't run this script."
    printf "Please report this to https://github.com/nextcloud/vm/issues/"
    exit 1
fi
```

And only do:

```
# T&M Hansson IT AB © - 2020, https://www.hanssonit.se/

# shellcheck disable=2034,2059
true
# shellcheck source=fetch_lib.sh
source /var/scripts/fetch_lib.sh
```

In the startup script we use something totally different, so leave that out of `fetch_lib` from that one and do it everywhere else as the logic for the startup script only happens there.

Reply:
Also, why do we want to cache the lib file? Don't we always want to use the latest and greatest?

Reply:
okay, my idea was to also cover these lines, so I guess this is where our solutions diverged 😅

Reply:
> Also, why do we want to cache the lib file? Don't we always want to use the latest and greatest?

I though about if it would make sense to cache the lib for 30min?
I am not sure if that really makes sense, though.

Reply:
So am I right, that you just want the following code in fetch_lib:
```
#!/bin/bash
# shellcheck disable=2034,2059
true
# see https://github.com/koalaman/shellcheck/wiki/Directive

# shellcheck source=lib.sh
source <(curl -sL https://raw.githubusercontent.com/nextcloud/vm/lib.sh) &>/dev/null
```

Reply:
> # T&M Hansson IT AB © - 2020, https://www.hanssonit.se/
> 
> # If we have internet, then use the latest variables from the lib remote file
> if printf "Testing internet connection..." && ping github.com -c 2
> then
> # shellcheck disable=2034,2059
> true
> SCRIPT_NAME="Server configuration"
> # shellcheck source=lib.sh
> . <(curl -sL https://raw.githubusercontent.com/nextcloud/vm/master/lib.sh)
> # Use local lib file in case there is no internet connection
> elif [ -f /var/scripts/lib.sh ]
> then
> # shellcheck disable=2034,2059
> true
> # shellcheck source=lib.sh
> source /var/scripts/lib.sh
> else
>     printf "You don't seem to have a working internet connection, and /var/scripts/lib.sh is missing so you can't run this script."
>     printf "Please report this to https://github.com/nextcloud/vm/issues/"
>     exit 1
> fi

This is exactly what we do in the startup script, no?

Reply:
Sry, I apparently don't get it, yet.